### PR TITLE
Add feature for allocating resident to rooms

### LINKED ISSUE #95: As a user, I can see a list of unallocated residents
... so that I can allocate them

### LINKED ISSUE #96: As a user, I can allocate a resident to a room
... so that he has a room.

### PR DESCRIPTION
**Pull request by colintkn**
### What this does ### 
This PR allows:
* allocation of resident to room
* deallocation of resident from room

Closes #95 
Closes #96 

### How to test ###
Launch SunRez.
`clear`
`radd n/Jane Doe p/98765432 e/johnd@example.com y/1`
`radd n/Jill Doe p/98765432 e/johnd@example.com y/1`
`radd n/Jo Doe p/98765432 e/johnd@example.com y/1`
`oadd r/12-125 t/CORRIDOR_AC g/SHN`
`alloc n/Jane Doe r/12-125`
_Room allocation field of Jane changes from `Room unallocated` to `12-125`_
_Occupancy field of the room changes from `N` to `Y`_
`redit 1 e/ jane@gmail.com `
_The resident has been allocated to a room. Please deallocate the resident before editing._
`rdel 1`
_The resident has been allocated to a room. Please deallocate the resident before deletion._
`oedit 1 t/ SUITE_AC`
_The room has already been allocated to another resident. Please deallocate the room before editing._
`odel 1`
_The room has been allocated to another resident. Please deallocate the room before deletion._
`rulist`
_Jane Doe should be filtered from the list_ 
`rlist`
`dealloc 2`
_The resident has not been allocated._
`dealloc 1`
_Room allocation field of Jane changes from  `12-125` to 'Room unallocated'_
_Occupancy field of the room changes from `Y` to `N`_

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/CommandResult.java [hunk 1]
```diff
@@ -21,6 +24,18 @@
...
     private final boolean exit;
 
+    /**
+     * Constructs a {@code CommandResult} with the specified fields.
+     */
+    public CommandResult(String feedbackToUser, boolean isResidentCommand, boolean isRoomCommand,
+            boolean isResidentRoomCommand, boolean showHelp, boolean exit) {
+        this.feedbackToUser = requireNonNull(feedbackToUser);
+        this.isResidentCommand = isResidentCommand;
+        this.isRoomCommand = isRoomCommand;
```
Comment (stein414) [inline src/main/java/seedu/address/logic/commands/CommandResult.java:34]: Maybe a enum like "CommandType" would make more sense as it is somewhat overcrowding the constructor. The check could be against the enum value instead if needed

File: src/main/java/seedu/address/logic/commands/CommandResult.java [hunk 2]
```diff
@@ -41,13 +59,67 @@ public CommandResult(String feedbackToUser, boolean showHelp, boolean exit) {
...
+        return new CommandResult(this.feedbackToUser, true, false, false, this.showHelp, this.exit);
+    }
+
+    /**
+     * Indicates if this CommandResult was the result of a Resident command
+     *
+     * @return CommandResult object indicating the result of the executed command
+     */
+    public boolean isResidentCommand() {
+        return this.isResidentCommand;
```
Comment (stein414) [inline src/main/java/seedu/address/logic/commands/CommandResult.java:84]: The naming seems ambiguous

File: src/main/java/seedu/address/model/residentroom/UniqueResidentRoomList.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    public void setResidentRoom(ResidentRoom target, ResidentRoom editedResidentRoom) {
+        requireAllNonNull(target, editedResidentRoom);
+
+        int index = internalList.indexOf(target);
+        if (index == -1) {
+            throw new ResidentNotFoundException();
+        }
+
+        if (!target.isSameResidentRoom(editedResidentRoom) && containsEitherResidentRoom(editedResidentRoom)) {
+            throw new DuplicateResidentException();
```
Comment (stein414) [inline src/main/java/seedu/address/model/residentroom/UniqueResidentRoomList.java:78]: Perhaps use `DuplicateResidentRoomException` instead

**Review by DrWala — APPROVED** [review]
Smoke test showed the feature working as expected. LGTM.

File: src/main/java/seedu/address/model/residentroom/UniqueResidentRoomList.java [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+import javafx.collections.FXCollections;
+import javafx.collections.ObservableList;
+import seedu.address.model.resident.Resident;
+import seedu.address.model.resident.exceptions.DuplicateResidentException;
+import seedu.address.model.resident.exceptions.ResidentNotFoundException;
+import seedu.address.model.residentroom.exceptions.DuplicateResidentRoomException;
+import seedu.address.model.residentroom.exceptions.ResidentRoomNotFoundException;
+
+/**
+ * A list of residents that enforces uniqueness between its elements and does not allow nulls.
```
Comment (stein414) [inline src/main/java/seedu/address/model/residentroom/UniqueResidentRoomList.java:18]: Perhaps change the resident references in this file to resident rooms (both in comment and code)

File: src/main/java/seedu/address/model/residentroom/UniqueResidentRoomList.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+        if (!target.isSameResidentRoom(editedResidentRoom) && containsEitherResidentRoom(editedResidentRoom)) {
+            throw new DuplicateResidentException();
+        }
+
+        internalList.set(index, editedResidentRoom);
+    }
+
+    /**
+     * Removes the equivalent resident from the list.
```
Comment (stein414) [inline src/main/java/seedu/address/model/residentroom/UniqueResidentRoomList.java:85]: resident -> resident room

File: src/main/java/seedu/address/storage/JsonAdaptedResidentRoom.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import seedu.address.commons.exceptions.IllegalValueException;
+import seedu.address.model.resident.Name;
+import seedu.address.model.resident.Resident;
+import seedu.address.model.residentroom.ResidentRoom;
+import seedu.address.model.room.RoomNumber;
+
+/**
+ * Jackson-friendly version of {@link Resident}.
```
Comment (stein414) [inline src/main/java/seedu/address/storage/JsonAdaptedResidentRoom.java:13]: Resident -> ResidentRoom

File: src/main/java/seedu/address/storage/JsonAdaptedResidentRoom.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    /**
+     * Converts a given {@code Resident} into this class for Jackson use.
+     */
+    public JsonAdaptedResidentRoom(ResidentRoom source) {
+        name = source.getName().fullName;
+        roomNumber = source.getRoomNumber().toString();
+    }
+
+    /**
+     * Converts this Jackson-friendly adapted resident object into the model's {@code Resident} object.
```
Comment (stein414) [inline src/main/java/seedu/address/storage/JsonAdaptedResidentRoom.java:40]: Resident -> ResidentRoom

**Review by stein414 — CHANGES_REQUESTED** [review]
A few references to residents are still there when it probably should be resident room. Other than that it seems fine.

**Review by benedictkhoomw — CHANGES_REQUESTED** [review]
There is an uncaught `IndexOutOfBoundsException` when trying to deallocate if the involved resident is not currently listed.

### How to reproduce ###
Launch SunRez and enter the following commands
`radd n/Jane Doe p/98765432 e/johnd@example.com y/1`
`radd n/Jill Doe p/98765432 e/johnd@example.com y/1`
`radd n/Jo Doe p/98765432 e/johnd@example.com y/1`
`oadd r/12-125 t/CORRIDOR_AC g/SHN`
`alloc n/Jane Doe r/12-125`
`rulist` (Jane Doe should be filtered from the list)
`dealloc n/Jane Doe r/12-125` (This should trigger the exception)

**Pull request merged**